clang/include/clang/Basic/arm_neon.td

```diff
@@ -1959,9 +2064,12 @@ multiclass VCMLA_ROTS<string type, string lanety, string laneqty> {
 
     let isLaneQ = 1 in  {
       // vcmla{ROT}_laneq
+      // ACLE specifies that the fp16 vcmla_#ROT_laneq variant has an immedaite range of 0 <= lane <= 1.
+      // fp16 is the only variant for which these two differ.
+      // https://developer.arm.com/documentation/ihi0073/latest/ 
+      defvar getlanety = !if(!eq(type, "h"), lanety, laneqty);
       def : SOpInst<"vcmla" # ROT # "_laneq", "...QI", type,  Op<(call "vcmla" # ROT, $p0, $p1,
-              (bitcast $p0, (dup_typed lanety, (call "vget_lane", (bitcast laneqty, $p2), $p3))))>>;
-
+                (bitcast $p0, (dup_typed lanety, (call "vget_lane", (bitcast getlanety, $p2), $p3))))>>;
```

Comment:
I am not sure I quite follow what is happening her. Could you elaborate on the purpose of getlanety ?

Reply:
The `vcmla` intrinsics are instantiated for each base type ([f16, f32, f64])  under this `VCMLA_ROTS` multiclass. For example:

`defm VCMLA_F32        : VCMLA_ROTS<"f", "uint64x1_t", "uint64x2_t">;`

The first argument is a vector with the number of lanes accessible by the `lane` argument in  [ vcmlaq_lane](https://developer.arm.com/architectures/instruction-sets/intrinsics/#q=vcmlaq_lane)  and [vcmla_lane](https://developer.arm.com/architectures/instruction-sets/intrinsics/#q=vcmla_lane) type inrinsics. The second argument is that for [vcmlaq_laneq](https://developer.arm.com/architectures/instruction-sets/intrinsics/#q=vcmlaq_laneq) and [vcmla_laneq](https://developer.arm.com/architectures/instruction-sets/intrinsics/#q=vcmla_laneq) type intrinsics . This is done to give a vector of the correct number of elements to the call to `vget` in the instruction definition, so that the range of the immediate is correctly bounded (the range of this immediate is such that it can access some of the lower half of the vector only, which differs from traditional `_lane` type intrinsics where we access an arbitrary lane).

There is a problem with this approach for the `f16`/`h` base type; the range of the immediate in `vcmlaq_laneq_f16` is different to that for `vcmla_laneq_f16` (` 0<=lane<=3` and `0<=lane<=1` respectively). This is how the intrinsics are instantiated:

`defm VCMLA_FP16  : VCMLA_ROTS<"h", "uint32x2_t", "uint32x4_t">;`

The simplest fix I found for this case was to add the `getlanety` conditional which will select `uint32x2_t` for `vcmla_laneq_f16` and `uin32x4_t` for `vcmlaq_laneq_f16`, giving the correct range for the immediates.

Reply:
I did take a look into  [ACLE release](https://github.com/ARM-software/acle/releases/tag/r2024Q2) and I think the website you linked might have a mistake there. Looking at advsimd-2023Q2 I see `vcmla[_rot]_laneq_f16` defined like this :
<img width="725" alt="image" src="https://github.com/user-attachments/assets/2a446a5c-a3bf-40fc-9950-6caee95d81a2">

It also makes sense for all of them to have the same imm. range, as the range is only dependant on size and datatype of the 3 operand and not on the sizes and datatypes of the other operands.

Reply:
If we look at what this intrinsic compiles to, [FCMLA](https://developer.arm.com/documentation/ddi0596/2021-03/SIMD-FP-Instructions/FCMLA--by-element---Floating-point-Complex-Multiply-Accumulate--by-element--?lang=en), it looks like the reduced range (`0<=lane<=1`) is due to the fact that we must have `H=='0'` for this instruction to be defined when we use the `'01'` size specifier and `Q='0'` (what is required for this variant). This means that only bit-field `L` can be used to encode `<index>`.

Thanks to @rsandifo-arm for pointing this out.

Reply:
I am not sure I follow what you trying to say there. I think, it might be because your link to FCMLA is for the old spec and there may be some differences in instruction spec compared to the latest one. But if you look at the most recent spec for this [instruction](https://developer.arm.com/documentation/ddi0602/2024-06/SVE-Instructions/FCMLA--indexed---Floating-point-complex-multiply-add-by-indexed-values-with-rotate-), it says quite clearly that immediate range for half-precision variant is from 0 to 3

Reply:
> [I]f you look at the most recent spec for this [instruction](https://developer.arm.com/documentation/ddi0602/2024-06/SVE-Instructions/FCMLA--indexed---Floating-point-complex-multiply-add-by-indexed-values-with-rotate-), it says quite clearly that immediate range for half-precision variant is from 0 to 3

That link is to the SVE instruction though, rather than the Advanced SIMD one.  The range for the 8H Advanced SIMD instruction is [0, 3], but the range for the 4H one is [0, 1].

Reply:
Apologies got confused there. Now I see what you are talking about. But to me it seems that whoever wrote this part of ACLE still wanted to allow lane to be from 0 to 3 and got around this limitation of the instruction by using DUP . So if we want to be consistent with ACLE we should still allow this wider range. Or do you think we should go and propose changing ACLE for this variant?

Reply:
No, I agree that the ACLE expects an index of [2,3] to be emulated using DUP.